### PR TITLE
Integrate expression data heatmap track feature

### PR DESCRIPTION
**Pull request by fedde-s**
This ports the heatmap code from pull request #1625 to the context of the current release candidate, fixing remaining issues with the implementation. ~~It is currently not yet functioning correctly, I believe because it still indexes its columns by sample or patient in calls to `Oncoprint.prototype.setTrackData()` as used to be the norm before the datamanager reorganisation in pull request #1493. In addition, it is still lacking a legend for the colours.~~

### ~~Screenshot~~
Screenshot of the heatmap with data aggregated to patient level

### ~~Screenshot (with current issues)~~

Screenshot of the misaligned heatmap tracks lacking a legend

### CODE REVIEW

**Comment by fedde-s** [comment]
@adamabeshouse, could you help adjust the code so that the columns line up with the pre-existing tracks again? I touched the functions that I believe need to be adapted in commit thehyve/cbioportal@dedeb330ba95c7fc90c4a17f30ca33a184076c54, but I think you know the code better and you might see what needs to be done much quicker.

**Comment by fedde-s** [comment]
In addition, you may know where to add a legend for the tracks.

**Comment by jjgao** [comment]
@fedde-s: thanks! I'd like to release this soon (in 1-2 weeks) together with MD Anderson's NGS-Heatmap pull request.

File: portal/src/main/webapp/js/api/cbioportal-datamanager.js [hunk 1]
```diff
@@ -1099,6 +1118,61 @@ window.initDatamanager = function (genetic_profile_ids, oql_query, cancer_study_
...
+	    ).then(function (client_sample_data, sample_to_patient_map) {
+		var interim_data = {};
+		for (var i = 0; i < genes.length; i++) {
+		    var gene = genes[i].toUpperCase();
+		    interim_data[gene] = {};
+		    for (var j = 0; j < sample_ids.length; j++) {
+			var id = sample_ids[j];
+			interim_data[gene][id] = {};
+			interim_data[gene][id]["hugo_gene_symbol"] = gene;
+			interim_data[gene][id][sample_or_patient] = (sample_or_patient === "patient" ? sample_to_patient_map[id] : id);
```
Comment (fedde-s) [inline portal/src/main/webapp/js/api/cbioportal-datamanager.js]: @adamabeshouse, what did the Oncoprint previously do when `setTrackData()` was given a list with multiple data for the same patient?

Comment (adamabeshouse) [inline portal/src/main/webapp/js/api/cbioportal-datamanager.js]: The way we do it now in the oncoprint is compile a list of data associated with a sample (or patient, depending on whether you want sample or patient data), and then create an "oncoprint datum" where the display parameters are computed from all the associated data. You would probably want to do something similar here - aggregate the data in a list and then go through all the data and compute the average or something like that.

**Comment by adamabeshouse** [comment]
image

**Comment by jjgao** [comment]
@fedde-s @adamabeshouse what's the status of this pull request?

**Comment by adamabeshouse** [comment]
@jjgao I've implemented the features they asked me to, I think there's just one issue remaining which @fedde-s has resolved but not yet pushed up.

**Comment by jjgao** [comment]
Thanks @adamabeshouse.

@fedde-s could you update? I'd like to merge soon. Thanks.

**Comment by fedde-s** [comment]
@jjgao: will do! Ran into a small delay testing the fix while troubleshooting that the current rc branch is no longer compatible with the version of Maven in Debian stable, but now that I upgraded I can probably test and push the commit tomorrow.
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by fedde-s** [comment]
I fixed the issue, and resolved remaining issues @pieterlukasse and I noticed when reviewing @adamabeshouse's fix. @adamabeshouse, shall I rebase on the current rc branch to remove merge commits and maybe squash fixup commits before final review?

**Comment by adamabeshouse** [comment]
@fedde-s Sounds good, either rebase or just merge and patch if rebasing is too messy.

**Comment by fedde-s** [comment]
@adamabeshouse, do you have more to add to the review of this pull request? And @jjgao, do you know of anyone else who should review it?

**Comment by jjgao** [comment]
@schultzn any comments?

http://bit.ly/2gmcmjJ

File: portal/src/main/webapp/js/src/oncoprint/webgl/setup.js [hunk 2]
```diff
@@ -791,6 +873,39 @@ window.CreateCBioPortalOncoprintWithToolbar = function (ctr_selector, toolbar_se
...
+			'label': heatmap_data_by_line[i].hugo_gene_symbol,
+			'target_group': 2,
+			//'sortCmpFn': function(d1, d2) {return 0;},
+			'removable': true,
+			'description': heatmap_data_by_line[i].hugo_gene_symbol,
+		    };
+		    var new_hm_id = oncoprint.addTracks([track_params])[0];
+		    hm_ids.push(new_hm_id);
+		    State.heatmap_tracks[i] = new_hm_id;
+		    // ???: wouldn't this always be State.heatmap_track[0]? <fedde-s 2016-10-06>
```
Comment (adamabeshouse) [inline portal/src/main/webapp/js/src/oncoprint/webgl/setup.js]: Remove comment, but I agree we don't need `State.first_heatmap_track` - replace the `if` test with a check for the length of heatmap_tracks, and replace `State.first_heatmap_track` with `State.heatmap_tracks[0]`

File: portal/src/main/webapp/js/src/oncoprint/webgl/setup.js [hunk 3]
```diff
@@ -1162,45 +1277,76 @@ window.CreateCBioPortalOncoprintWithToolbar = function (ctr_selector, toolbar_se
...
+		// heatmap not applicable, return a resolved Promise
+		return $.when();
 	    }
-	    oncoprint.scrollTo(0);
-	});
-	
-	return populate_data_promise;
-    });
+	}).then(function fetchClinicalAttributes() {
+	    console.log("in initOncoprint, fetching clinical attributes");
```
Comment (adamabeshouse) [inline portal/src/main/webapp/js/src/oncoprint/webgl/setup.js]: This is not using promises correctly. The `return` statements from the nested callbacks will not reach the outside the way this intends it to. `return State.clinical_attributes_fetched.promise()` should be at the bottom of `fetchClinicalAttributes` and the nested `return` statements should be replaced with simply `State.clinical_attributes_fetched.resolve()` and `.reject()`

**Comment by jjgao** [comment]
[ ] fix color schema: the values in the middle should be either white or black.
[ ] remove space between rows
[ ] always fill the space between columns

File: portal/src/main/webapp/js/src/oncoprint/webgl/setup.js (hunk 3 above)
Comment (adamabeshouse) [inline portal/src/main/webapp/js/src/oncoprint/webgl/setup.js]: I addressed this in https://github.com/cBioPortal/cbioportal/pull/1847/commits/86ece36645fdaa1f43266ea7c5e228788ac344a7

File: portal/src/main/webapp/js/api/cbioportal-datamanager.js [hunk 4]
```diff
@@ -953,6 +953,31 @@ window.initDatamanager = function (genetic_profile_ids, oql_query, cancer_study_
...
 	    return this.genetic_profile_ids;
 	},
+	/**
+	 * Selects the genetic profile most suitable for heatmap rendering.
+	 *
+	 * @returns {Promise<?string>} promise resolving with the
+	 * stable ID of the profile, or with null if no applicable
+	 * profiles are in use.
+	 */
+	'getDefaultHeatmapProfile': function () {
```
Comment (adamabeshouse) [inline portal/src/main/webapp/js/api/cbioportal-datamanager.js]: I addressed this in https://github.com/cBioPortal/cbioportal/pull/1847/commits/86ece36645fdaa1f43266ea7c5e228788ac344a7

**Comment by adamabeshouse** [comment]
image
image

**Comment by adamabeshouse** [comment]
@fedde-s Regarding the commits, it may also make sense to go back to before my merge and cherry-pick my non-merge commits?

**Comment by fedde-s** [comment]
@adamabeshouse: absolutely, that's basically what I was planning to do. It is conceptually what rebasing does behind the scenes, although [this webcomic](https://xkcd.com/1597/) is appropriate here.

File: portal/src/main/webapp/js/api/cbioportal-datamanager.js [hunk 5]
```diff
@@ -1099,6 +1124,113 @@ window.initDatamanager = function (genetic_profile_ids, oql_query, cancer_study_
...
+	 */
+	/**
+	 * Fetches Oncoprint heatmap data per gene track for the given profile per sample or patient.
+	 *
+	 * @param {string} genetic_profile_id - identifier of the genetic profile to query
+	 * @param {string[]} genes - list of gene symbols for which to return data
+	 * @param {string} sample_or_patient - whether to return data per sample or aggregate to patient level
+	 * @returns {Promise<OncoprintHeatmapTrackData[]>}
+	 */
+	'getHeatmapData': function (genetic_profile_id, genes, sample_or_patient) {
```
Comment (pieterlukasse) [inline portal/src/main/webapp/js/api/cbioportal-datamanager.js]: we will pick this up and split in 2 functions: one for patients, one for samples

File: portal/src/main/webapp/js/src/oncoprint/webgl/setup.js (hunk 2 above)
Comment (pieterlukasse) [inline portal/src/main/webapp/js/src/oncoprint/webgl/setup.js]: ok

File: portal/src/main/webapp/js/src/oncoprint/webgl/setup.js (hunk 3 above)
Comment (pieterlukasse) [inline portal/src/main/webapp/js/src/oncoprint/webgl/setup.js]: small todo: remove `return` in fail function

**Comment by fedde-s** [comment]
Removed the problematic merge, by rebasing the two commits depending on it.

**Comment by adamabeshouse** [comment]
How does this look?
image

**Comment by jjgao** [comment]
@adamabeshouse it looks great. Anything else left?

File: portal/src/main/webapp/js/src/oncoprint/webgl/setup.js (hunk 2 above)
Comment (fedde-s) [inline portal/src/main/webapp/js/src/oncoprint/webgl/setup.js]: Done.

**Comment by pieterlukasse** [comment]
@adamabeshouse awesome! Will the user also be able to switch color schemes via the UI toolbar? Would be great if that is possible.

**Comment by jjgao** [comment]
@pieterlukasse @adamabeshouse let's not worry about switching color scheme for now. In the future, we can design a colorblind-friendly scheme for the whole oncoprint and allow users to switch.

File: portal/src/main/webapp/js/api/cbioportal-datamanager.js [hunk 6]
```diff
@@ -960,24 +965,35 @@ window.initDatamanager = function (genetic_profile_ids, oql_query, cancer_study_
...
 		}
-		return null;
-	    }).promise();
-	},
+		if (mrna_profile_id !== null) {
+		    fetch_promise.resolve(mrna_profile_id);
+		} else if (prot_profile_id !== null) {
+		    fetch_promise.resolve(prot_profile_id);
+		} else {
+		    fetch_promise.reject();
```
Comment (fedde-s) [inline portal/src/main/webapp/js/api/cbioportal-datamanager.js]: This is an error: the function is documented to resolve with `null` if data can be retrieved but is not applicable, that is what the functions that call it expect. If this function does not resolve, rendering of the Oncoprint will not continue.

File: portal/src/main/webapp/js/api/cbioportal-datamanager.js [hunk 7]
```diff
@@ -891,7 +891,12 @@ window.initDatamanager = function (genetic_profile_ids, oql_query, cancer_study_
 		fetcher(this, fetch_promise);
 	    }
 	    fetch_promise.then(function (data) {
-		def.resolve(deepCopyObject(data));
+		var ret = data;
+		if ([(typeof ""), (typeof 0), (typeof true), (typeof undefined)].indexOf(typeof data) === -1
```
Comment (adamabeshouse) [inline portal/src/main/webapp/js/api/cbioportal-datamanager.js]: done

File: portal/src/main/webapp/js/src/oncoprint/webgl/setup.js (hunk 3 above)
Comment (adamabeshouse) [inline portal/src/main/webapp/js/src/oncoprint/webgl/setup.js]: oops yes, good catch, thanks. done.

File: portal/src/main/webapp/js/src/oncoprint/webgl/setup.js [hunk 8]
```diff
@@ -1162,45 +1280,77 @@ window.CreateCBioPortalOncoprintWithToolbar = function (ctr_selector, toolbar_se
...
+	    QuerySession.getClinicalAttributes()
+	    .then(function (attrs) {
+		console.log("in initOncoprint, clinical attributes fetched");
+		State.unused_clinical_attributes = attrs;
+		State.clinical_attributes_fetched.resolve();
+	    }).fail(function () {
+		return State.clinical_attributes_fetched.reject();
```
Comment (adamabeshouse) [inline portal/src/main/webapp/js/src/oncoprint/webgl/setup.js]: thanks, yep.

**Comment by adamabeshouse** [comment]
What's going on with the documentation above `getHeatmapData`? Do we want to keep all that?

**Comment by fedde-s** [comment]
@adamabeshouse: If you don't, do you have any suggestion where else to document what the method actually returns?

**Comment by adamabeshouse** [comment]
@fedde-s You're right, it's worthwhile to keep there. At some point we'll transition all of this to Typescript.

**Comment by jjgao** [comment]
@fedde-s could you please try to disable heatmap for protein expression data?

@fedde-s @adamabeshouse any other remaining items?

File: portal/src/main/webapp/js/src/oncoprint/webgl/setup.js [hunk 9]
```diff
@@ -791,6 +873,39 @@ window.CreateCBioPortalOncoprintWithToolbar = function (ctr_selector, toolbar_se
...
+		var hm_ids = [];
+		for (var i = 0; i < heatmap_data_by_line.length; i++) {
+		    var track_params = {
+			'rule_set_params': {
+			    'type': 'gradient',
+			    'value_key': 'profile_data',
+			    'value_range': [-3, 3],
+			    'colormap': 'inferno',
+			    //'color_range': ['rgba(0,0,255,1)', 'rgba(255,0,0,1)'],
+			    'null_color': 'rgba(211,211,211,1)'
```
Comment (adamabeshouse) [inline portal/src/main/webapp/js/src/oncoprint/webgl/setup.js]: done

**Comment by adamabeshouse** [comment]
@jjgao LGTM

**Comment by fedde-s** [comment]
Thanks, @adamabeshouse. Sorry it took me a while, I was busy some issues in my local test instances after loading the newly accepted seed database.

**Pull request merged**